### ISSUE TITLE
PR 8 — `Predictor` + `Provider` protocol + `IncrementalLabelsWriter`

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#516]
**Parent epic:** #508
**Prerequisites:** #512
**Blocks:** #518

---

## Goal

Ship Layer 3 — the orchestration layer that handles batching, I/O providers, filtering, tracking, progress reporting, and `sio.Labels` construction. This is the biggest PR in the stack; it replaces the current 52-param `run_inference()` driver.

After this PR, the complete user-facing pipeline works:

```python
predictor = Predictor.from_model_paths(["models/topdown/"])
labels = predictor.predict("video.mp4")
labels.save("predictions.slp")
```

## Three-tier memory model

Per design doc §4.9:

| Method | Peak memory | Use for |
|---|---|---|
| `predict(source)` | O(N_frames) — full `sio.Labels` in RAM | Notebooks, short clips |
| `predict_streaming(source)` | O(tracker_window) + O(queue_size) | Real-time / custom loops |
| `predict_to_file(source, path)` | O(write_interval) — flushes to `.slp` | **Production default for long videos** |

`predict()` = `list(predict_streaming())`; `predict_to_file()` = `predict_streaming()` + `IncrementalLabelsWriter`.

### `FilterConfig` — single source of truth for all filtering

**Design decision (design doc §4B.4):** `Outputs` has no filter methods. All filtering — keypoint-level AND instance-level — lives here. `FilterPipeline` is the executor.

```python
# sleap_nn/inference/predictor.py  (config portion)

@attrs.define
class FilterConfig:
    """Post-inference filtering configuration.

    Single source of truth for every filter applied between the
    InferenceLayer's raw Outputs and the final sio.Labels.
    """

    # ── Keypoint-level (NaN-out individual points) ───────────────────
    min_peak_value: float = 0.0

    # ── Instance-level: score / count thresholds ─────────────────────
    min_instance_score: float = 0.0
    min_mean_node_score: float = 0.0
    min_visible_nodes: int = 0
    min_visible_node_fraction: float = 0.0

    # ── Instance-level: overlap NMS ──────────────────────────────────
    # Both filters can be enabled simultaneously: IoU catches duplicates of
    # the same animal in odd poses (near-identical bbox, very different
    # keypoints); OKS catches similar-pose duplicates of distinct close
    # animals. Setting both to non-None runs IoU first (cheaper) then OKS
    # on survivors. Set to None to disable.
    # See: https://github.com/talmolab/sleap/discussions/2653
    overlapping_iou_threshold: Optional[float] = None
    overlapping_oks_threshold: Optional[float] = None

    # ── Final cap: keep top-N by instance score ──────────────────────
    # Runs LAST, after all quality filters. This is intentional: if two
    # duplicate detections of one animal both outscore a real instance on
    # a different animal, we want the duplicate culled by overlap NMS first
    # and only THEN cap to N. The pre-refactor code applied this cap
    # before the quality filters, which inverted the desired behavior.
    # See: https://github.com/talmolab/sleap/discussions/2664
    # NOTE: distinct from `CentroidLayer.max_instances`, which is a
    # stage-1 compute-saving cap on the centroid model output. This filter
    # is a post-pipeline quality cap.
    max_instances_per_frame: Optional[int] = None
```

### `FilterPipeline` — the executor

Callable. One-off classmethod. Fixed documented order.

```python
@attrs.define
class FilterPipeline:
    """Apply a FilterConfig to Outputs in a fixed, documented order.

    Order matters — this class exists so the order is in ONE place, not
    scattered across _make_labeled_frames_from_generator() like today.
    """
    config: FilterConfig

    def __call__(self, outputs: Outputs) -> Outputs:
        """Shorthand: `pipeline(outputs)` instead of `pipeline.apply(outputs)`."""
        return self.apply(outputs)

    def apply(self, outputs: Outputs) -> Outputs:
        """Canonical order: cheap → expensive; each step narrows candidates.

          1. min_peak_value         — keypoint-level, O(B·I·N)
          2. Node-count filters     — min_visible_nodes, min_visible_node_fraction
          3. Score filters          — min_mean_node_score, min_instance_score
          4. Overlap NMS (IoU)      — O(B·I²) on bbox math; runs first because
                                      cheaper and catches the obvious duplicates
          5. Overlap NMS (OKS)      — O(B·I²·N); runs on the IoU survivors
          6. max_instances_per_frame — sort by score, take top-N (runs LAST so
                                      the cap operates on the cleaned set, not
                                      raw detections)
        """
        from sleap_nn.inference.ops.filters import (
            threshold_keypoints, filter_by_node_count, filter_by_confidence,
            greedy_nms_iou, greedy_nms_oks, cap_top_n_by_score,
        )
        if self.config.min_peak_value > 0:
            outputs = threshold_keypoints(outputs, self.config.min_peak_value)
        if self.config.min_visible_nodes or self.config.min_visible_node_fraction:
            outputs = filter_by_node_count(
                outputs,
                min_visible_nodes=self.config.min_visible_nodes,
                min_visible_node_fraction=self.config.min_visible_node_fraction,
            )
        if self.config.min_instance_score or self.config.min_mean_node_score:
            outputs = filter_by_confidence(
                outputs,
                min_instance_score=self.config.min_instance_score,
                min_mean_node_score=self.config.min_mean_node_score,
            )
        if self.config.overlapping_iou_threshold is not None:
            outputs = greedy_nms_iou(
                outputs, threshold=self.config.overlapping_iou_threshold,
            )
        if self.config.overlapping_oks_threshold is not None:
            outputs = greedy_nms_oks(
                outputs, threshold=self.config.overlapping_oks_threshold,
            )
        if self.config.max_instances_per_frame is not None:
            outputs = cap_top_n_by_score(
                outputs, n=self.config.max_instances_per_frame,
            )
        return outputs

    @classmethod
    def run(cls, outputs: Outputs, config: FilterConfig) -> Outputs:
        """One-off convenience: `FilterPipeline.run(outputs, config)`."""
        return cls(config=config)(outputs)
```

### New ops in `ops/filters.py`

**`threshold_keypoints`** — keypoint-level (not instance-level), previously called `min_peak_value` inside `Outputs.filter_by_confidence`. Now a standalone op for consistency:

```python
def threshold_keypoints(outputs: Outputs, min_peak_value: float) -> Outputs:
    """NaN-out keypoints whose peak_value < threshold. Does not drop instances."""
    mask = outputs.pred_peak_values < min_peak_value
    new_keypoints = outputs.pred_keypoints.clone()
    new_keypoints[mask] = float("nan")
    new_peak_values = outputs.pred_peak_values.clone()
    new_peak_values[mask] = 0.0
    return attrs.evolve(
        outputs, pred_keypoints=new_keypoints, pred_peak_values=new_peak_values,
    )
```

**`cap_top_n_by_score`** — final post-filter cap. Sorts by `instance_scores` descending and keeps top-N per frame; the rest are marked invalid (`instance_valid=False`) so downstream `to_instances()` skips them.

```python
def cap_top_n_by_score(outputs: Outputs, n: int) -> Outputs:
    """Keep the top-n instances per frame by instance_scores; mark rest invalid.

    Pre-condition: prior filters have already NaN-ed out failed instances,
    which is reflected in instance_scores (NaN sorts to bottom). This op runs
    LAST in the pipeline so that quality filters get first say.
    """
    # (B, I) → indices of top-n per row, padded with -1 for invalid slots
    scores = outputs.instance_scores  # (B, I)
    # NaN scores sort to the end with descending=True only if we replace with -inf
    safe_scores = torch.where(torch.isnan(scores), torch.full_like(scores, -float("inf")), scores)
    topn_idx = safe_scores.topk(min(n, scores.shape[1]), dim=1).indices  # (B, n)
    keep_mask = torch.zeros_like(scores, dtype=torch.bool)
    keep_mask.scatter_(1, topn_idx, True)
    # Combine with existing instance_valid (don't resurrect already-filtered slots)
    if outputs.instance_valid is not None:
        keep_mask = keep_mask & outputs.instance_valid
    return attrs.evolve(outputs, instance_valid=keep_mask)
```

### `Provider` protocol

```python
# sleap_nn/inference/providers.py

from typing import Iterator, Protocol, Tuple, runtime_checkable
import numpy as np
import torch


@runtime_checkable
class Provider(Protocol):
    """Source of inference batches.

    Concrete implementations: VideoProvider, LabelsProvider, NumpyProvider.
    """

    def __len__(self) -> int:
        """Total number of frames (for progress)."""
        ...

    def iter_batches(self, batch_size: int) -> Iterator["Batch"]:
        """Yield batches of raw frames with metadata."""
        ...


@attrs.define
class Batch:
    """One batch from a provider."""
    images: torch.Tensor           # (B, C, H, W), uint8
    frame_indices: torch.Tensor    # (B,)
    video_indices: torch.Tensor    # (B,)
    # GT-only, for use_gt_centroids mode:
    instances: Optional[torch.Tensor] = None
```

Three concrete providers wrapping existing `VideoReader` / `LabelsReader` + a new `NumpyProvider` for in-memory arrays.

### `Predictor`

```python
# sleap_nn/inference/predictor.py

@attrs.define
class Predictor:
    """High-level inference orchestrator.

    Composes an InferenceLayer with batching, filtering, tracking,
    I/O providers, progress reporting, and sio.Labels construction.
    """

    layer: InferenceLayer
    batch_size: int = 4
    filter_config: FilterConfig = attrs.Factory(FilterConfig)
    tracker: Optional[Tracker] = None
    paf_workers: int = 0                       # #517 plumbs this through
    reset_tracker_between_videos: bool = True
    progress: "ProgressReporter" = attrs.Factory("RichProgressReporter")

    @classmethod
    def from_model_paths(
        cls,
        model_paths: List[Union[str, Path]],
        device: str = "auto",
        use_compile: bool = False,
        tracker_config: Optional[Dict[str, Any]] = None,
        backbone_ckpt_path: Optional[str] = None,
        head_ckpt_path: Optional[str] = None,
        **kwargs,
    ) -> "Predictor":
        """Auto-detect model type + backend and construct the right layer."""
        ...

    # ─── Three-tier memory model ───────────────────────────────────

    def predict(
        self,
        source: Union[np.ndarray, torch.Tensor, sio.Video, sio.Labels, str, Path,
                      List[Union[str, Path, sio.Video]]],
        as_labels: bool = True,
        frames: Optional[List[int]] = None,
    ) -> Union[sio.Labels, List[Outputs]]:
        """Run inference, accumulating all results.

        Accepts single source or list. Multi-video: results merged into one Labels.
        Warns if source has > MAX_MEMORY_SAFE_FRAMES (10_000).
        """
        if self._is_large_source(source):
            import warnings
            warnings.warn(
                "Source has >10k frames; consider predict_to_file() to avoid OOM."
            )
        all_outputs = list(self.predict_streaming(source, frames=frames))
        return self._to_labels(all_outputs, source) if as_labels else all_outputs

    def predict_streaming(
        self,
        source,
        frames: Optional[List[int]] = None,
        progress: bool = True,
    ) -> Iterator[Outputs]:
        """Memory-efficient streaming.

        Per-frame cost: O(tracker_window) + O(queue_size).
        """
        provider = self._build_provider(source, frames=frames)
        filter_pipe = FilterPipeline(self.filter_config)

        for video_idx, video_provider in enumerate(self._iter_videos(provider)):
            if self.tracker and self.reset_tracker_between_videos and video_idx > 0:
                self.tracker.reset()

            for batch in video_provider.iter_batches(self.batch_size):
                outputs = self.layer.predict(batch.images)
                outputs = filter_pipe.apply(outputs)
                outputs.frame_indices = batch.frame_indices
                outputs.video_indices = batch.video_indices
                if self.tracker:
                    outputs = self.tracker.track(outputs)
                yield outputs.slim()

    def predict_to_file(
        self,
        source,
        output_path: Union[str, Path],
        frames: Optional[List[int]] = None,
        write_interval: int = 500,
        resume: bool = False,
    ) -> sio.Labels:
        """Stream predictions directly to .slp file.

        Memory: O(write_interval) regardless of video length.
        """
        writer = IncrementalLabelsWriter(
            output_path, skeleton=self.layer.skeleton,
            write_interval=write_interval, resume=resume,
        )
        for outputs in self.predict_streaming(source, frames=frames):
            writer.append(outputs)
        return writer.finalize()

    @classmethod
    def track_only(
        cls,
        labels: sio.Labels,
        tracker_config: Dict[str, Any],
        **kwargs,
    ) -> sio.Labels:
        """Re-track existing predictions without inference.

        Replaces the `run_inference(model_paths=None, tracking=True)` shortcut.
        """
        ...
```

### `IncrementalLabelsWriter`

```python
# sleap_nn/inference/streaming.py

@attrs.define
class IncrementalLabelsWriter:
    """Flushes outputs to .slp in batches to keep memory bounded.

    Per design doc §4.9:
    - Writes to <out>.inflight; atomic rename to <out> on clean exit
    - Idempotent flushes keyed by frame counter
    - resume=True continues from last committed frame
    - SIGINT → flush buffer, rename to <out>.partial, exit non-zero
    """

    output_path: Path
    skeleton: sio.Skeleton
    write_interval: int = 500
    resume: bool = False

    _buffer: List[Outputs] = attrs.field(factory=list, init=False)
    _frame_counter: int = attrs.field(default=0, init=False)
    _inflight_path: Path = attrs.field(init=False)

    def __attrs_post_init__(self):
        self._inflight_path = self.output_path.with_suffix(".slp.inflight")
        if self.resume:
            self._load_committed_frames()
        self._install_sigint_handler()

    def append(self, outputs: Outputs) -> None:
        self._buffer.append(outputs.slim())
        self._frame_counter += outputs.batch_size
        if len(self._buffer) * self.write_interval <= self._frame_counter:
            self._flush()

    def _flush(self) -> None:
        """Merge buffer into .inflight and clear buffer."""
        ...

    def finalize(self) -> sio.Labels:
        """Flush remaining buffer, rename .inflight → output_path, return Labels."""
        ...
```

### `ProgressReporter`

```python
# sleap_nn/inference/progress.py

class ProgressReporter(Protocol):
    def start(self, total: int) -> None: ...
    def advance(self, n: int = 1) -> None: ...
    def finish(self) -> None: ...


class RichProgressReporter:
    """Default terminal progress (preserves current RateColumn behavior)."""
    ...


class JsonProgressReporter:
    """JSON-per-line to stdout, throttled to 4 Hz (for --gui mode)."""
    ...
```

## Multi-video sequencing (per §3.6a)

- Videos processed sequentially (no interleaving)
- Tracker state resets at each video boundary by default
- `Outputs.video_indices` set per batch
- `IncrementalLabelsWriter` writes one merged `.slp` with `Labels.videos = [v1, v2, ...]`

## Tracking-only mode

```python
# Used when model_paths is None and tracking=True in the old run_inference
Predictor.track_only(existing_labels, tracker_config={
    "method": "hungarian", "window_size": 10, "scoring": "oks",
})
```

## Acceptance criteria

- [ ] All three prediction methods (`predict`, `predict_streaming`, `predict_to_file`) work end-to-end on a fixture video with every model type
- [ ] `NumpyProvider` accepts raw numpy arrays directly (covers the "real-time API" user story)
- [ ] `LabelsProvider` supports all six frame-selection flags from §3.6 (`--frames`, `--only_labeled_frames`, `--only_suggested_frames`, `--exclude_user_labeled`, `--only_predicted_frames`, `--no_empty_frames`)
- [ ] Multi-video support: `predict(["v1.mp4", "v2.mp4"])` returns one merged `Labels` with correct `video_indices`
- [ ] `predict_to_file()` with 500-frame fixture uses ≤ O(500) memory (verified with `tracemalloc`)
- [ ] Resume test: kill a `predict_to_file()` run midway, re-run with `resume=True`, output matches uninterrupted run
- [ ] SIGINT test: Ctrl-C during `predict_to_file()` produces `.partial` file, doesn't corrupt
- [ ] `JsonProgressReporter` output matches current `--gui` format (compare against golden JSON stream)
- [ ] `track_only()` replaces the current tracking-only shortcut
- [ ] `FilterPipeline` applies filters in documented order (add explicit test)
- [ ] `FilterPipeline` is callable (`pipeline(outputs)` works) and has `.run(outputs, config)` classmethod
- [ ] `FilterConfig` covers all filters (keypoint-level `min_peak_value` + instance-level + dual NMS + max-N cap) — no filter option exists elsewhere
- [ ] `ops/filters.py::threshold_keypoints` added for the keypoint-level NaN-out (moved from removed `Outputs.filter_by_confidence`)
- [ ] **Dual overlap NMS** — `FilterConfig` accepts `overlapping_iou_threshold` and `overlapping_oks_threshold` independently; setting both runs IoU first, then OKS on survivors. Test with a fixture where IoU-only and OKS-only each leave a duplicate that the combined pass removes. (Addresses [SLEAP #2653](https://github.com/talmolab/sleap/discussions/2653).)
- [ ] **`max_instances_per_frame` runs LAST** — `ops/filters.py::cap_top_n_by_score` added; pipeline test confirms cap is applied after overlap NMS, not before. Regression test: a fixture with two duplicate detections of animal A (scores 0.8, 0.75) plus one real detection of animal B (score 0.6) and `max_instances_per_frame=2` keeps animal A (deduped to one) + animal B, not both copies of A. (Addresses [SLEAP #2664](https://github.com/talmolab/sleap/discussions/2664).)
- [ ] **No conflict with `CentroidLayer.max_instances`** — both can be set; layer-level cap (stage-1 compute saver) is enforced inside `CentroidLayer`, filter-level cap (post-quality cap) inside `FilterPipeline`. Documented in `FilterConfig` docstring; test confirms they compose correctly when both are non-None.
- [ ] `Outputs` has no filter methods — verified by `grep` check in CI (no `def filter*` in `outputs.py`)
- [ ] `ruff check` + `black --check` clean

## Files touched

**Created:**
- `sleap_nn/inference/predictor.py`
- `sleap_nn/inference/providers.py`
- `sleap_nn/inference/streaming.py`
- `sleap_nn/inference/progress.py`
- `tests/inference/test_predictor.py`
- `tests/inference/test_providers.py`
- `tests/inference/test_streaming.py`
- `tests/inference/test_multi_video.py`

## Out of scope

- PAF worker pool (plumbed via `paf_workers` but implementation is #517)
- CLI wiring (#518)
- Deletion of old `predict.py::run_inference` (#519)
- Compat shims for old `TopDownPredictor` etc. names (#520)

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
